36978 | Fast image processor for DPT model

samrae7 opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

Add Fast Image Processor for DPT model

<!-- Remove if not applicable -->

Fixes #36978

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. [Link to issue](https://github.com/huggingface/transformers/issues/36978)
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @eustlb
- graph models: @clefourrier

Library:

- flax: @gante and @Rocketknight1
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @zach-huggingface and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @SunMarc @zach-huggingface
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @Rocketknight1
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

Commit:
chore: ran codegen script

Commit:
test: test_image_processor_properties

Commit:
test: test_image_processor_from_dict_with_kwargs

Commit:
test: wip - test_padding

Commit:
test: test_padding

Commit:
test: test_keep_aspect_ratio

Commit:
wip

Commit:
test

Commit:
test: wip

Commit:
test: wip

Commit:
test: test_call_segmentation_maps, wip

Commit:
chore: tidy up

Commit:
test: test_call_segmentation_maps

Commit:
fix: test_save_load_fast_slow

Commit:
test: reduce labels

Commit:
chore: make fixup

samrae7 commented on src/transformers/models/dpt/image_processing_dpt_fast.py:
```diff
@@ -0,0 +1,414 @@
+
+
+def get_output_image_size_for_ensure_multiple(
+    # TODO type should be torch.Tensor
```
^

Commit:
chore: rm comment

samrae7 commented on src/transformers/models/dpt/image_processing_dpt_fast.py:
```diff
@@ -0,0 +1,414 @@
+    return (new_height, new_width)
+
+
+class DPTFastImageProcessorKwargs(DefaultFastImageProcessorKwargs):
```
TODO: sort out documentation for these

Commit:
chore: tidy

Commit:
chore remove comment

samrae7 commented on src/transformers/models/dpt/image_processing_dpt_fast.py:
```diff
@@ -0,0 +1,400 @@
+            if do_center_crop:
+                stacked_images = self.center_crop(stacked_images, crop_size)
+            if do_pad:
+                stacked_images = torch.stack([self.pad_image(image, size_divisor) for image in stacked_images])
```
Could do with some guidance here:

Is it much preferable to make pad_image handle batched tensors, and so avoid a list comprehension like this?

Is that for speed, ie. to take advantage of pytorch perfromance?

If so I can make that change quite quickly

yonigozlan replied:
Hey @samrae7 , yes that would be great!

Commit:
refactor: no need to infer channel dimesnion

samrae7 commented on src/transformers/models/dpt/image_processing_dpt_fast.py:
```diff
@@ -0,0 +1,400 @@
+        data = {"pixel_values": processed_images}
+
+        if segmentation_maps is not None:
+            isList = False
```
TODO: factor thios out into a method, _preprocess_segmentation_maps

Commit:
refactor: encapsulate logic for preparing segmentation maps

Commit:
refactor: improve readability of segmentation_map preparation

Commit:
improvement: batched version of pad_image

Commit:
chore: fixup

Commit:
docs

Commit:
chore: make quality

Commit:
chore: remove unecessary comment

Commit:
fix: add SemanticSegmentationMixin

samrae7 commented:
@yonigozlan I've looked at the test failures in the pipeline and they seem unrelated to my change. The branch was updated from upstream main just before the tests ran. Any advice on what to do?

![Uploading Screenshot 2025-04-19 at 22.13.48.png…]()

samrae7 commented:
Update: I was on holiday for a week and can now pick this up again. There were apparently-unrelated tests failing in the pipeline. I will investigate

Commit:
feat: add post_process_depth_estimation to fast dpt image processor

Commit:
chore: fix formatting

yonigozlan commented on src/transformers/models/dpt/image_processing_dpt_fast.py:
```diff
@@ -0,0 +1,451 @@
+        Prepares segmentation maps for processing. Leverages _prepare_input_images,
+        hence to use that we need to add and remove a dimension which corresponds to the missing batch dimension.
+        """
+        isBatched = False
+        if isinstance(segmentation_maps, list):
+            isBatched = True
+            if not is_pil_image(segmentation_maps[0]) and segmentation_maps[0].ndim == 2:
+                segmentation_maps = [map.unsqueeze(0) for map in segmentation_maps]
+        elif not is_pil_image(segmentation_maps) and segmentation_maps.ndim == 2:
+            segmentation_maps = segmentation_maps.unsqueeze(0)
+
+        segmentation_maps = self._prepare_input_images(segmentation_maps)
+
+        if isBatched:
+            segmentation_maps = [map.squeeze(0) for map in segmentation_maps]
+        else:
+            segmentation_maps = segmentation_maps[0]
+        return segmentation_maps
```
Could you have a look at how this is handled in the newly merged BeitImageProcessorFast?

samrae7 replied:
Sure I'll do that, thanks 👍

samrae7 replied:
thanks for this suggestion. I've refactored based on the approach used in Beit Fast Processor

yonigozlan reviewed (commented):
Hey @samrae7 ! Thanks for iterating :). The Beit fast image processor was recently merged, and it seems quite similar to this one, could you have a look to try and uniformize this PR a bit based on it? Especially the handling of segmentation maps, as I'm not a fan of dragging them as ImageInput inside the `_preprocess` function

Commit:
remove max_height, max_width

Commit:
fix: better way of processin segmentation maps
- copied from Beit Fast processor

Commit:
chore: formatting + remove TODO

Commit:
chore: fixup styles

samrae7 commented on src/transformers/models/dpt/image_processing_dpt_fast.py:
```diff
@@ -0,0 +1,451 @@
+                input_data_format=ChannelDimension.FIRST,
+            )
+        elif size.max_height and size.max_width:
+            new_size = get_image_size_for_max_height_width(image.size()[-2:], size.max_height, size.max_width)
```
removed this

Commit:
chore: remove unecessary line break

yonigozlan reviewed (commented):
Thanks for iterating! I'd prefer having this be closer to Beit fast image processor, as the logic is almost the same

samrae7 commented on src/transformers/models/dpt/image_processing_dpt_fast.py:
```diff
@@ -0,0 +1,397 @@
+        data = {"pixel_values": processed_images}
+
+        if segmentation_maps is not None:
+            segmentation_maps = make_list_of_images(images=segmentation_maps, expected_ndims=2)
+            segmentation_maps = self._preprocess_segmentation_maps(
+                segmentation_maps=segmentation_maps,
+                return_tensors=return_tensors,
+                **kwargs,
+            )
```
ok will make this and the other changes asap. thanks

samrae7 replied:
done

Commit:
chore: core review suggestion to remove autodocstring

Commit:
fix: add do_reduce_labels logic + refactor
- refactor preprocess logic to make it consistent with other processors
- add missing reduce labels logic

Commit:
refactor: remove deprecated mixin

Commit:
chore: fixup

samrae7 commented on src/transformers/models/dpt/image_processing_dpt_fast.py:
```diff
@@ -0,0 +1,397 @@
+    size_divisor: Optional[int]
+    do_pad: Optional[bool]
+    keep_aspect_ratio: Optional[bool]
+    segmentation_maps: Optional[ImageInput] = (None,)
```
added now thanks for spotting that!

samrae7 commented on src/transformers/models/dpt/image_processing_dpt_fast.py:
```diff
@@ -0,0 +1,397 @@
+
+
+@auto_docstring
+class DPTImageProcessorFast(BaseImageProcessorFast, SemanticSegmentationMixin):
```
done. thanks

samrae7 committed:
Merge branch 'main' into 36978/dpt-fast-image

yonigozlan committed:
use modular for dpt + final nit changes

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into 36978/dpt-fast-image

yonigozlan committed:
fix style

yonigozlan reviewed (approved):
Hey @samrae7 ! Thanks a lot for working on this, looks great!
I made some final changes, mainly to use modular as a large part of the processing code was the same as Beit. Waiting for the CI then LGTM!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_37481). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
